subprojects/core/src/main/java/org/gradle/initialization/layout/BuildLayoutFactory.java

```diff
@@ -85,6 +89,21 @@ BuildLayout getLayoutFor(File currentDir, File stopAt) {
         return layout(currentDir, new File(currentDir, Settings.DEFAULT_SETTINGS_FILE));
     }
 
+    private File findSettingsFileInMasterSubdirectory(File root) {
+        File settingsFile = findExistingSettingsFileIn(new File(root, "master"));
+        if (settingsFile != null) {
+            deprecateMasterSubdirectoryForRootBuild();
+        }
+        return settingsFile;
+    }
+
+    private static void deprecateMasterSubdirectoryForRootBuild() {
```

Comment:
OK, so I think this deprecation is getting lost because `BuildLayoutFactory` is used by services and gets loaded very, very early.  We can just move where this is done though.

Instead of emitting the deprecation here, let's make `SettingsLocation` have a "master directory" flag that's set when we find the settings file in a master directory.

Then in `DefaultSettingsLoader`, let's check that flag and emit the deprecation then. By that point, it should be safe to emit deprecation messages.

If that does work, we should add a integ test for it.

Reply:
Thanks! This seems to have worked.